src/server/services/wms/qgswmsrenderer.cpp

```diff
@@ -644,131 +644,161 @@ namespace QgsWms
+      i = mWmsParameters.xAsInt();
+      j = mWmsParameters.yAsInt();
     }
-
-    //In case the output image is distorted (WIDTH/HEIGHT ratio not equal to BBOX width/height), I and J need to be adapted as well
-    int widthParam = mParameters.value( "WIDTH", "-1" ).toInt();
-    int heightParam = mParameters.value( "HEIGHT", "-1" ).toInt();
-    if ( ( i != -1 && j != -1 && widthParam != -1 && heightParam != -1 ) && ( widthParam != outputImage->width() || heightParam != outputImage->height() ) )
+    int width = mWmsParameters.widthAsInt();
+    int height = mWmsParameters.heightAsInt();
+    if ( ( i != -1 && j != -1 && width != 0 && height != 0 ) && ( width != outputImage->width() || height != outputImage->height() ) )
     {
-      i *= ( outputImage->width() / ( double )widthParam );
-      j *= ( outputImage->height() / ( double )heightParam );
+      i *= ( outputImage->width() / ( double )width );
+      j *= ( outputImage->height() / ( double )height );
     }
 
-    //Normally, I/J or X/Y are mandatory parameters.
-    //However, in order to make attribute only queries via the FILTER parameter, it is allowed to skip them if the FILTER parameter is there
-
+    // init search variables
     std::unique_ptr<QgsRectangle> featuresRect;
+    std::unique_ptr<QgsGeometry> filterGeom;
     std::unique_ptr<QgsPointXY> infoPoint;
 
-    if ( i == -1 || j == -1 )
-    {
-      if ( mParameters.contains( QStringLiteral( "FILTER" ) ) )
-      {
-        featuresRect.reset( new QgsRectangle() );
-      }
-      else if ( !filterGeom.get() )
-      {
-        throw QgsBadRequestException( QStringLiteral( "ParameterMissing" ),
-                                      QStringLiteral( "I/J parameters are required for GetFeatureInfo" ) );
-      }
-    }
-    else
+    if ( i != -1 && j != -1 )
     {
       infoPoint.reset( new QgsPointXY() );
       infoPointToMapCoordinates( i, j, infoPoint.get(), mapSettings );
     }
+    else if ( filtersDefined )
+    {
+      featuresRect.reset( new QgsRectangle() );
+    }
+    else if ( filterGeomDefined )
+    {
+      filterGeom.reset( new QgsGeometry( QgsGeometry::fromWkt( mWmsParameters.filterGeom() ) ) );
+    }
 
-    //get the layer registered in QgsMapLayerRegistry and apply possible filters
-    ( void )layerSet( layersList, stylesList, mapSettings.destinationCrs() );
-
-    //scoped pointer to restore all original layer filters (subsetStrings) when pointer goes out of scope
-    //there's LOTS of potential exit paths here, so we avoid having to restore the filters manually
-    std::unique_ptr< QgsOWSServerFilterRestorer > filterRestorer( new QgsOWSServerFilterRestorer( mAccessControl ) );
-    applyRequestedLayerFilters( layersList, mapSettings, filterRestorer->originalFilters() );
-
-#ifdef HAVE_SERVER_PYTHON_PLUGINS
-    applyAccessControlLayersFilters( layersList, filterRestorer->originalFilters() );
-#endif
+    QDomDocument result;
```

Comment:
Do you think we could build the *QDomDocument result* in another method? It would shrink the *getFeatureInfo* which is a little big.